### PR TITLE
Fix memory leak, add experimental setAll_P()

### LINKED ISSUE #14: Changing the pointer from malloc() memory leak
The pointer to the heap is changed. The pointer to the memory in the heap that was allocted with malloc() should not be changed.

In the function setAll(), the pointer to the heap is changed when doing: `_digitalValues = digitalValues;`.
After that, the allocated memory in the heap can no longer be used.
When the user had the data on the stack, the _digitalValues-pointer could be pointing to a memory location that is no longer valid or is used for something else.
This can be avoided with memcpy() to store the new data in the heap.

For example, storing the new data in the heap, and after that, using the values from the heap for shiftOut.
```
    memcpy( _digitalValues, digitalValues, numberOfShiftRegisters);   // dest, src, size
    
    for (byte = _numberOfShiftRegisters - 1; byte >= 0; byte--) {
        shiftOut(_serialDataPin, _clockPin, MSBFIRST, _digitalValues[byte]);
    }

```

When that bug is fixed, you can add new things:

- It would be nice to add a destructor that releases the memory with free().
- A overloaded function could be added to support PROGMEM data for the setAll() function.

Note: It is possible to use a template to create a constructor runtime with an array of a certain size.

### PR DESCRIPTION
**Pull request by Koepel**
Merging resolves https://github.com/Simsso/ShiftRegister74HC595/issues/14

### CODE REVIEW

**Comment by Simsso** [comment]
Thank you, I really appreciate you took the time and fixed the code. I'll review and test during the next days and merge afterwards. Once again, thank you. Such high quality contributions are rare! 🥇

File: src/ShiftRegister74HC595.cpp [hunk 1]
```diff
@@ -14,46 +14,58 @@ ShiftRegister74HC595::ShiftRegister74HC595(int numberOfShiftRegisters, int seria
...
+// For example with:
+//     const uint8_t myFlashData[] PROGMEM = { 0x0F, 0x81 };
+#ifdef __AVR__
+void setAll_P(const uint8_t * digitalValuesProgmem)
+{
+    PGM_VOID_P p = reinterpret_cast<PGM_VOID_P>(digitalValuesProgmem);
+    memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
+    updateRegisters();
+}
+#endif
```
Comment (Simsso) [inline src/ShiftRegister74HC595.cpp:68]: I have tried to compile with the Arduino IDE version 1.8.9 on Mac, for Arduino Uno and am getting the following error:

```
src/ShiftRegister74HC595.cpp: In function 'void setAll_P(const uint8_t*)':
src/ShiftRegister74HC595.cpp:63:15: error: '_digitalValues' was not declared in this scope
     memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
               ^
src/ShiftRegister74HC595.cpp:63:34: error: '_numberOfShiftRegisters' was not declared in this scope
     memcpy_P( _digitalValues, p, _numberOfShiftRegisters);
                                  ^
src/ShiftRegister74HC595.cpp:64:21: error: 'updateRegisters' was not declared in this scope
     updateRegisters();
                     ^
src/ShiftRegister74HC595.cpp: In member function 'void ShiftRegister74HC595::updateRegisters()':
src/ShiftRegister74HC595.cpp:90:55: error: 'digitalValues' was not declared in this scope
         shiftOut(_serialDataPin, _clockPin, MSBFIRST, digitalValues[i]);
                                                       ^
exit status 1
Error compiling for board Arduino/Genuino Uno.
```

File: src/ShiftRegister74HC595.cpp [hunk 2]
```diff
@@ -71,9 +83,15 @@ void ShiftRegister74HC595::set(int pin, uint8_t value)
 }
 
 // Updates the shift register pins to the stored output values.
+// This is the function that actually writes data into the shift registers of the 74HC595
 void ShiftRegister74HC595::updateRegisters()
 {
-	 setAll(_digitalValues);
+    for (int i = _numberOfShiftRegisters - 1; i >= 0; i--) {
+        shiftOut(_serialDataPin, _clockPin, MSBFIRST, digitalValues[i]);
```
Comment (Simsso) [inline src/ShiftRegister74HC595.cpp]: Little typo, should be `_digitalValues`, here.

**Review by Simsso — CHANGES_REQUESTED** [review]

File: src/ShiftRegister74HC595.cpp (hunk 2 above)
Comment (Simsso) [reply src/ShiftRegister74HC595.cpp]: I have taken care of that one in this commit: https://github.com/Simsso/ShiftRegister74HC595/pull/15/commits/c496675fee4e48a53c5de7297f8fd9d09e0ca41b

**Comment by Koepel** [comment]
O, I'm sorry. Updating the shift registers with the data from the heap is with `_digitalValues` of course.
Is there something I should do ?

**Comment by Simsso** [comment]
Yes, concerning [this error](https://github.com/Simsso/ShiftRegister74HC595/pull/15#discussion_r268982245), do you know what the cause is? Where did you get it to compile? If not I'd suggest removing the experimental parts for now and would merge the PR with the other improvements only.

**Comment by Koepel** [comment]
I really messed up big time. The function getAll_P did not know the variables of the class, because it was not part of the class. Changed `getAll_P(` to `ShiftRegister74HC595::getAll_P(`.
This time I removed my test versions and carefully checked that my fork is okay.

The Arduino Zero and Arduino MKR allow to use the keyword `PROGMEM` and also the function `memcpy_P()`. That makes the source compatible with the Arduino Uno. Even the ESP8266 libraries allow to use `PROGMEM`. Perhaps the `#ifdef __AVR__` and `#endif` can be removed, but I don't know if it is guaranteed accepted for all Arduino boards. It is easier when the user copies the PROGMEM data into a local buffer and just use `setAll()`.

**Review by Simsso — APPROVED** [review]

**Comment by Simsso** [comment]
Thank you once again. I've done some testing with actual hardware and the changes look very good. Merged ✅

**Comment by Koepel** [comment]
Sorry once again. I've should have tested the final pull request instead of a test version on my computer.

You could add a destructor with a `free(this->data)` to your `ShiftRegister-PWM-Library` as well.

**Comment by Simsso** [comment]
I think we're freeing the allocated memory here, don't we: https://github.com/Simsso/ShiftRegister74HC595/blob/master/src/ShiftRegister74HC595.cpp#L44

**Comment by Koepel** [comment]
Yes, I meant for your "PWM" library.

**Pull request merged**